src/dodal/devices/oav/pin_tip_detection/oav_with_pin_tip_detection.py

```diff
@@ -0,0 +1,214 @@
+import asyncio
+import time
+from collections import OrderedDict
+from typing import Optional, Tuple, Type, TypeVar
+
+import numpy as np
+from bluesky.protocols import Descriptor
+from numpy.typing import NDArray
+from ophyd.v2.core import Device, Readable, Reading, SimSignalBackend
+from ophyd.v2.epics import SignalR, SignalRW, epics_signal_r
+
+from dodal.devices.oav.pin_tip_detection.pin_tip_detect_utils import (
+    ARRAY_PROCESSING_FUNCTIONS_MAP,
+    ArrayProcessingFunctions,
+    MxSampleDetect,
+    ScanDirections,
+)
+from dodal.log import LOGGER
+
+T = TypeVar("T")
+
+
+def _create_soft_signal(name: str, datatype: Type[T]) -> SignalRW[T]:
+    return SignalRW(
+        SimSignalBackend(
+            datatype, f"sim://oav_with_pin_tip_detection_soft_params:{name}"
+        )
+    )
+
+
+class PinTipDetection(Readable, Device):
+    """
+    A device which will read a single frame from an on-axis view and use that frame
+    to calculate the pin-tip offset (in pixels) of that frame.
+
+    Used for pin tip centring workflow.
+
+    Note that if the sample is off-screen, this class will return the centre as the "edge"
+    of the image.
+    """
+
+    def __init__(self, prefix: str, name: str = ""):
+        self._prefix: str = prefix
+        self._name = name
+
+        self.array_data: SignalR[NDArray[np.uint8]] = epics_signal_r(
+            NDArray[np.uint8], f"pva://{prefix}PVA:ARRAY"
+        )
+
+        self.oav_width: SignalR[int] = epics_signal_r(
+            int, f"{prefix}PVA:ArraySize1_RBV"
+        )
+        self.oav_height: SignalR[int] = epics_signal_r(
+            int, f"{prefix}PVA:ArraySize2_RBV"
+        )
+
+        # Soft parameters for pin-tip detection.
+        self.timeout: SignalRW[float] = _create_soft_signal("timeout", float)
+        self.preprocess: SignalRW[int] = _create_soft_signal("preprocess", int)
+        self.preprocess_ksize: SignalRW[int] = _create_soft_signal(
+            "preprocess_ksize", int
+        )
+        self.preprocess_iterations: SignalRW[int] = _create_soft_signal(
+            "preprocess_iterations", int
+        )
+        self.canny_upper: SignalRW[int] = _create_soft_signal("canny_upper", int)
+        self.canny_lower: SignalRW[int] = _create_soft_signal("canny_lower", int)
+        self.close_ksize: SignalRW[int] = _create_soft_signal("close_ksize", int)
+        self.close_iterations: SignalRW[int] = _create_soft_signal(
+            "close_iterations", int
+        )
+        self.scan_direction: SignalRW[int] = _create_soft_signal("scan_direction", int)
+        self.min_tip_height: SignalRW[int] = _create_soft_signal("min_tip_height", int)
```

Comment:
It was done this way to mirror the an existing dodal device, where these parameters are PVs in an ADPython plugin - the idea being that the interfaces should be similar from a plan's perspective so that we can swap between the implementations if necessary. At least in the short term.

These parameters will/might be changed from the plan level `bps.set(image_detection.canny_upper, 500)` etc. Is there a cleaner way to allow `bps.set(...)` or similar that doesn't involve using the sim backend? 

Conceptually it's a slight abuse of the "sim" backend as a generic "soft" backend I guess, but that's the cleanest way I found currently...

Reply:
Interesting. That makes sense; I can speak to tom indepth about this when he gets back, because we should find a way to support this in ophyd or explain why we aren't properly.